### PR TITLE
RavenDB-13110 - Sharding - external and pull replication integration

### PR DESCRIPTION
**Pull request by shiranshalom**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-13110

### Additional description

This draft PR contains  an abstraction for `IncomingReplicationHandler` & `OutgoingReplicationHandler`. 

### Type of change

- New feature

### How risky is the change?

- Not relevant

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs [hunk 1]
```diff
@@ -29,7 +29,8 @@ public enum DatabaseItemType
         CompareExchangeTombstones = 1 << 17,
         TimeSeries = 1 << 18,
 
-        ReplicationHubCertificates = 1 << 19
+        ReplicationHubCertificates = 1 << 19,
+        AllDatabaseItems = Documents | RevisionDocuments | Tombstones | Conflicts | Attachments | CounterGroups | TimeSeries
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs]: move somewhere else, this is exposed in the client API

File: src/Raven.Server/Documents/Replication/Incoming/AbstractIncomingReplicationHandler.cs [hunk 2]
```diff
@@ -0,0 +1,668 @@
...
+                From = ConnectionInfo,
+                ExceptionMessage = exceptionMessage
+            });
+        }
+
+        protected abstract void EnsureNotDeleted(string nodeTag);
+
+        protected abstract void OnDocumentsReceived();
+
+        protected abstract void OnAttachmentStreamsReceives(int attachmentStreamCount);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Replication/Incoming/AbstractIncomingReplicationHandler.cs]: Received

File: src/Raven.Server/Documents/Replication/Incoming/AbstractIncomingReplicationHandler.cs [hunk 3]
```diff
@@ -0,0 +1,668 @@
...
+
+        protected abstract void OnFailed(Exception exception);
+
+        protected abstract Task HandleBatch(TOperationContext context, IncomingReplicationHandler.DataForReplicationCommand batch, long lastEtag);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Replication/Incoming/AbstractIncomingReplicationHandler.cs]: HandleBatchAsync

File: src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs [hunk 4]
```diff
@@ -920,100 +316,29 @@ public LiveReplicationPerformanceCollector.ReplicationPerformanceType GetReplica
                 : LiveReplicationPerformanceCollector.ReplicationPerformanceType.IncomingExternal;
         }
 
-        public bool IsDisposed => _disposeOnce.Disposed;
+        protected override void OnAttachmentStreamsReceives(int attachmentStreamCount) => AttachmentStreamsReceived?.Invoke(this, attachmentStreamCount);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs]: we are calling those methods in Client API, InvokeXYZ, e.g. InvokeOnAttachmentStreamsReceived, maybe lets stick to that

File: src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs [hunk 5]
```diff
@@ -920,100 +316,29 @@ public LiveReplicationPerformanceCollector.ReplicationPerformanceType GetReplica
...
 
-        public void Dispose()
-        {
-            _disposeOnce.Dispose();
-        }
+        protected override void OnFailed(Exception exception) => Failed?.Invoke(this, exception);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs]: InvokeOnFailed

File: src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs [hunk 6]
```diff
@@ -920,100 +316,29 @@ public LiveReplicationPerformanceCollector.ReplicationPerformanceType GetReplica
...
-                {
-                    // can't do anything about it...
-                }
-            }
-        }
-
-        protected void OnFailed(Exception exception, IncomingReplicationHandler instance) => Failed?.Invoke(instance, exception);
-
-        protected void OnDocumentsReceived(IncomingReplicationHandler instance) => DocumentsReceived?.Invoke(instance);
+        protected override void OnDocumentsReceived() => DocumentsReceived?.Invoke(this);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs]: InvokeOnDocumentsReceived

File: src/Raven.Server/Documents/Replication/InterruptibleRead.cs [hunk 7]
```diff
@@ -9,21 +9,23 @@
 
 namespace Raven.Server.Documents.Replication
 {
-    public class InterruptibleRead : IDisposable
+    public class InterruptibleRead<TPool, TContext> : IDisposable
+    where TPool : JsonContextPoolBase<TContext>
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Replication/InterruptibleRead.cs]: TContextPool

File: src/Raven.Server/Documents/Replication/InterruptibleRead.cs [hunk 8]
```diff
@@ -9,21 +9,23 @@
...
+    where TContext : JsonOperationContext
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Replication/InterruptibleRead.cs]: TOperationContext

File: src/Raven.Server/Documents/Replication/Outgoing/AbstractOutgoingReplicationHandler.cs [hunk 9]
```diff
@@ -0,0 +1,804 @@
...
+            {
+                OccurredAt = SystemTime.UtcNow,
+                Direction = direction,
+                To = Destination,
+                IsExternal = GetType() != typeof(OutgoingInternalReplicationHandler),
+                ExceptionMessage = exceptionMessage
+            });
+        }
+
+        protected abstract void InitiatePullReplicationAsSink(TcpConnectionHeaderMessage.SupportedFeatures socketResultSupportedFeatures, X509Certificate2 certificate);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Replication/Outgoing/AbstractOutgoingReplicationHandler.cs:740]: can we move those abstract methods to the same place, now you have them in two places in this file. It is more readable if they are one after another.

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+        {
+            _parent = parent;
+            _replicationQueue = replicationQueue;
+            _attachmentStreamsTempFile = new StreamsTempFile("ShardedReplication" + Guid.NewGuid(), false);
+        }
+
+        protected override void EnsureNotDeleted(string nodeTag) => _parent.EnsureNotDeleted(_parent.Server.NodeTag);
+
+        protected override void OnFailed(Exception exception)
+        {
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs:38]: add DevevlopmentHelper, it looks like we are doing bunch of stuff in the Failed in regular handler

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+        }
+
+        protected override void HandleHeartbeatMessage(TransactionOperationContext jsonOperationContext, BlittableJsonReaderObject blittableJsonReaderObject)
+        {
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs:43]: add DevelopmentHelper here with appropriate description, we need to think what to do here. In regular handler this is a place where we are tackling the optimization issue when e.g. when we have master-master replication

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+                        if (shardAttachments.ContainsKey(attachment.Key))
+                        {
+                            var attachmentStream = new AttachmentReplicationItem
+                            {
+                                Type = ReplicationBatchItem.ReplicationItemType.AttachmentStream,
+                                Base64Hash = attachment.Key,
+                                Stream = new MemoryStream()
+                            };
+
+                            _attachmentStreamsTempFile._file.InnerStream.CopyTo(attachmentStream.Stream);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs:109]: DevHelper. We can optimize this and reduce number of copying

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+            return _replicationQueue.SendToShardCompletion.WaitAsync();
+        }
+
+        private void PrepareReplicationDataForShards(TransactionOperationContext context, IncomingReplicationHandler.DataForReplicationCommand dataForReplicationCommand)
+        {
+            var dictionary = new Dictionary<int, List<ReplicationBatchItem>>();
+            for (var shard = 0; shard < _parent.Context.ShardCount; shard++)
+                dictionary[shard] = new List<ReplicationBatchItem>();
+
+            foreach (var item in dataForReplicationCommand.ReplicatedItems)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs:74]: DevHelper. Optimization possibility: instead of iterating over materialized batch, we can do it while reading from the stream

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+            {
+                case AttachmentReplicationItem:
+                case AttachmentTombstoneReplicationItem:
+                case CounterReplicationItem:
+                case DocumentReplicationItem:
+                case TimeSeriesDeletedRangeItem:
+                case TimeSeriesReplicationItem:
+                    return _parent.Context.GetShardNumber(context, _documentInfoHelper.GetDocumentId(item));
+                case RevisionTombstoneReplicationItem:
+                    throw new NotSupportedInShardingException("TODO: implement for sharding"); // revision tombstones doesn't contain any info about the doc. The id here is the change-vector of the deleted revision
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs:147]: add also DevHelper one line above.

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+                                continue;
+                            }
+
+                            using (_stats.Network.Start())
+                            {
+                                MissingAttachmentsInLastBatch = false;
+
+                                var didWork = SendDocumentsBatch(context, items, _stats.Network);
+
+                                if (MissingAttachmentsInLastBatch) // TODO
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs]: DevHelper

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+                    }
+                }
+            }
+        }
+
+        private bool SendDocumentsBatch(TransactionOperationContext context, List<ReplicationBatchItem> items, OutgoingReplicationStatsScope stats)
+        {
+            try
+            {
+                //TODO: think how to unify this with the ReplicationDocumentSenderBase.SendDocumentsBatch
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs]: DevHelper

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+        protected override DynamicJsonValue GetInitialHandshakeRequest()
+        {
+            var initialRequest = base.GetInitialHandshakeRequest();
+
+            initialRequest[nameof(ReplicationLatestEtagRequest.SourceDatabaseId)] = _parent.SourceDatabaseId;
+            return initialRequest;
+        }
+
+        protected override void AddAlertOnFailureToReachOtherSide(string msg, Exception e)
+        {
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs:181]: DevHelper

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+        }
+
+        protected override void InitiatePullReplicationAsSink(TcpConnectionHeaderMessage.SupportedFeatures socketResultSupportedFeatures, X509Certificate2 certificate)
+        {
+            throw new NotSupportedInShardingException();
+        }
+
+        protected override void AssertDatabaseNotDisposed()
+        {
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs:191]: you can assert the context?

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+        }
+
+        protected override X509Certificate2 GetCertificateForReplication(ReplicationNode destination, out TcpConnectionHeaderMessage.AuthorizationInfo authorizationInfo)
+        {
+            return _parent.GetCertificateForReplication(Destination, out authorizationInfo);
+        }
+
+        protected override void OnBeforeDispose()
+        {
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs:202]: DevHelper

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+        }
+
+        protected override void OnSuccessfulTwoWaysCommunication()
+        {
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs:207]: DevHelper

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs [hunk 21]
```diff
@@ -0,0 +1,210 @@
...
+        }
+
+        protected override void OnFailed(Exception e)
+        {
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs:212]: DevHelper

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedTcpConnectionInfoHandlerForDatabase.cs [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+using Raven.Server.Extensions;
+using Raven.Server.Routing;
+using Sparrow.Json;
+
+namespace Raven.Server.Documents.Sharding.Handlers
+{
+    public class ShardedTcpConnectionInfoHandlerForDatabase : ShardedDatabaseRequestHandler
+    {
+        [RavenShardedAction("/databases/*/info/tcp", "GET")]
+        public async Task Get()
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedTcpConnectionInfoHandlerForDatabase.cs]: can remove, we have this. Just rebase.

File: src/Raven.Server/Documents/Sharding/ShardedDatabaseContext.ReplicationLoader.cs [hunk 23]
```diff
@@ -0,0 +1,409 @@
...
+using Sparrow.Json.Sync;
+using Sparrow.Logging;
+using Sparrow.Server.Json.Sync;
+using Voron;
+
+namespace Raven.Server.Documents.Sharding
+{
+    public partial class ShardedDatabaseContext
+    {
+        public readonly ShardedReplicationContext ReplicationContext;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/ShardedDatabaseContext.ReplicationLoader.cs]: Just rename property to Replication

File: src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs (hunk 1 above)
Comment (shiranshalom) [reply src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs]: done

File: src/Raven.Server/Documents/Replication/Incoming/AbstractIncomingReplicationHandler.cs (hunk 2 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/Incoming/AbstractIncomingReplicationHandler.cs]: done

File: src/Raven.Server/Documents/Replication/Incoming/AbstractIncomingReplicationHandler.cs (hunk 3 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/Incoming/AbstractIncomingReplicationHandler.cs]: done

File: src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs (hunk 4 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs]: done

File: src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs (hunk 5 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs]: done

File: src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs (hunk 6 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs]: done

File: src/Raven.Server/Documents/Replication/InterruptibleRead.cs (hunk 7 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/InterruptibleRead.cs]: done

File: src/Raven.Server/Documents/Replication/InterruptibleRead.cs (hunk 8 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/InterruptibleRead.cs]: done

File: src/Raven.Server/Documents/Replication/Outgoing/AbstractOutgoingReplicationHandler.cs (hunk 9 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/Outgoing/AbstractOutgoingReplicationHandler.cs:740]: done

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs (hunk 10 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs:38]: done

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs (hunk 11 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs:43]: done

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs (hunk 12 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs:109]: done

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs (hunk 13 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs:74]: done

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs (hunk 14 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs:147]: done

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs (hunk 15 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs]: done

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs (hunk 16 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs]: done

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs (hunk 17 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs:181]: done

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs (hunk 18 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs:191]: done

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs (hunk 19 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs:202]: done

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs (hunk 20 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs:207]: done

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs (hunk 21 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs:212]: done

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedTcpConnectionInfoHandlerForDatabase.cs (hunk 22 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/ShardedTcpConnectionInfoHandlerForDatabase.cs]: done

File: src/Raven.Server/Documents/Sharding/ShardedDatabaseContext.ReplicationLoader.cs (hunk 23 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/ShardedDatabaseContext.ReplicationLoader.cs]: done

File: src/Raven.Server/Documents/Replication/InterruptibleRead.cs [hunk 24]
```diff
@@ -61,18 +62,22 @@ public InterruptibleRead(DocumentsContextPool contextPool, Stream stream)
                 return ReturnAndClearValue();
             }
 
-            if (_previousWait.TryGetValue(interrupt, out Task<Task> task) == false)
+            if (interrupt != null)
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/InterruptibleRead.cs:65]: why this change?

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs [hunk 25]
```diff
@@ -0,0 +1,221 @@
+﻿extern alias NGC;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs]: remove this

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs [hunk 26]
```diff
@@ -0,0 +1,221 @@
...
+using Sparrow.Json.Parsing;
+using Sparrow.Logging;
+using Sparrow.Utils;
+using Voron;
+
+namespace Raven.Server.Documents.Sharding.Handlers
+{
+    public class ShardedOutgoingReplicationHandler : AbstractOutgoingReplicationHandler<TransactionContextPool, TransactionOperationContext>
+    {
+        private readonly int _shard;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs]: rename this to `_shardNumber`

**Comment by karmeli87** [comment]
please rebase and check failing tests

File: src/Raven.Server/Documents/Replication/InterruptibleRead.cs (hunk 24 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/InterruptibleRead.cs:65]: In the current state, `_replicationFromAnotherSource` (on `AbstractIncomingReplicationHandler.ReceiveReplicationBatches`) is null for sharding...

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs (hunk 25 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs]: done

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs (hunk 26 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs]: done

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs [hunk 27]
```diff
@@ -922,100 +316,29 @@ public LiveReplicationPerformanceCollector.ReplicationPerformanceType GetReplica
...
+
             public MergedDocumentReplicationCommand(DataForReplicationCommand replicationInfo, long lastEtag)
             {
                 _replicationInfo = replicationInfo;
                 _lastEtag = lastEtag;
             }
 
+            public MergedDocumentReplicationCommand(DataForReplicationCommand replicationInfo, long lastEtag, ReplicationLatestEtagRequest.ReplicationType replicationType) : this(replicationInfo, lastEtag)
+            {
+                _replicationType = replicationType;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs:339]: do we need that? can we override ` SaveSourceEtag` if needed?

File: src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs [hunk 28]
```diff
@@ -1361,7 +685,8 @@ protected override long ExecuteCmd(DocumentsOperationContext context)
             protected virtual void SaveSourceEtag(DocumentsOperationContext context)
             {
                 context.LastReplicationEtagFrom ??= new Dictionary<string, long>();
-                context.LastReplicationEtagFrom[_replicationInfo.SourceDatabaseId] = _lastEtag;
+                if (_replicationType != ReplicationLatestEtagRequest.ReplicationType.Sharded)
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs:688]: remove this?

File: src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs [hunk 29]
```diff
@@ -1440,6 +761,105 @@ public override TransactionOperationsMerger.IReplayableCommandDto<TransactionOpe
         }
     }
 
+    public unsafe class IncomingReplicationAllocator : IDisposable
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs:764]: move to new file

File: src/Raven.Server/Documents/Replication/InterruptibleRead.cs (hunk 24 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/Replication/InterruptibleRead.cs:65]: I would like to review this again at later point. please add here DevelopmentHelper..

**Review by karmeli87 — APPROVED** [review]

File: src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs (hunk 27 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs:339]: done

File: src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs (hunk 28 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs:688]: done (I added a check for `_replicationInfo.SourceDatabaseId != null `because we might get null for Sharded)

File: src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs (hunk 29 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs:764]: done
